corehq/apps/dump_reload/sql/load.py

```diff
@@ -58,6 +54,8 @@ def terminate_workers(dbalias_to_workerqueue):
 
         num_aliases = len(settings.DATABASES)
         manager = mp.Manager()
+        Executor = ProcessPoolExecutor if os.cpu_count() >= num_aliases \
+            else ThreadPoolExecutor
```

Comment:
(How) do you know if it is better to use `ThreadPoolExecutor` with low CPU count? I would expect processes would always either be better or worse, depending on the type of task being done.

Reply:
There is more than one factor to consider, but the general heuristic I've been applying:

* Task uses a lot of CPU: Use processes
* Task waits for other services: Use threads

Database tasks are a bit of both. I watched my CPU load as I tested, and the processes didn't max out the cores, but they spread noticeable load across them, so I leaned to using processes rather than threads.

As you know, processes are assigned to cores. So if the number of processes is greater than the number of cores, you've got cores sharing processes, which is kind of like using very expensive threads. If you're running this code on a bare-metal system with 8 cores and 5 DB aliases, processes are perfect. But if you're on a VM with a single core and 9 DB aliases, threads will be better.

It could be that the overhead for processes is not a huge consideration because this is a long-running task on machines that ought to have the RAM, but the docs in the library persuaded me that I was thinking along the right lines:

From Python 3.8 ProcessPoolExecutor docstring:

> max_workers: ... If None or not given then as many
> worker processes will be created as the machine has processors.

From Python 3.8 ThreadPoolExecutor comments:

>         if max_workers is None:
>             # We use cpu_count + 4 ...
>             # But we limit it to 32 to avoid consuming surprisingly
>             # large resource on many core machine.

(Prior to 3.8, ProcessPoolExecutor and ThreadPoolExecutor both used "cpu_count + 4". The fact that ProcessPoolExecuter changed, in my mind, added weight to my belief.)

But I could definitely be wrong about this, so if you think that the nature of the task outweighs the hardware its running on, I'd be happy to switch to a single approach. Also, if you think I should lean towards threads instead of processes, I'm also open to being persuaded.

Reply:
> if you're on a VM with a single core and 9 DB aliases, threads will be better.

To me this seems not necessarily a safe assumption, but it probably depends a lot on the workload. Threads are limited by the GIL, and therefore task switches are managed by the Python runtime. I've heard rumors that this can be inefficient in some cases (watch the first 4 minutes of [this talk](https://pyvideo.org/pycon-us-2010/pycon-2010--understanding-the-python-gil---82.html) for a demonstration of that). Maybe this has changed in modern Python versions?

Processes on the other hand are managed by the OS, which I assume has more robust not to mention far more stress-tested task switching mechanics than the Python runtime.

I admit what I'm saying here is pretty hand wavy, so if your experience indicates that what you're doing is good then feel free to ignore me.

Reply:
Oh man, those first four minutes are awesome. I will watch the rest later.

Yes, I know a little bit about the GIL. And you're right, it does get more complicated, because, if I remember correctly, `threading` threads are also created using the OS, but `asyncio` threads are not? (Maybe? I'm not sure. Maybe Py3 no longer uses the OS for any threads. I'd google it, but my electricity is going to get cut soon. But definitely worth exploring a bit more.)

Reply:
I should also add, after watching the entire talk, that near the end he shows a slide that says

> Don't use this talk to justify not using threads

Which is pretty much exactly what I did here. So I abused that talk, sorry for that. Also, there are hints in the talk that the situation has probably gotten much better in Python 3 (since Python 3.2 at least). But I think it still makes sense to measure the performance of your program and verify that it meets expectations.

Reply:
So, after managing to see how Travis is running TestSQLDumpLoadShardedModels in test_sql_dump_load.py, I discovered the issue: 

When Travis uses processes, the test passes. But when Travis uses threads, and adds a domain_migration_flags.domainmigrationprogress record to the "default" DB, the model's `save()` call hangs. It does not throw an exception.

I haven't dug any deeper. My first guess is that it's related to locking, but I'm not sure why the problem doesn't happen when using processes.

But maybe I should accept this as the "performance measure" I'm looking for :upside_down_face: and come back to the "why" some other time.